File: smartreminderssave/src/main/java/jgappsandgames/smartreminderssave/tags/TagManager.java
package jgappsandgames.smartreminderssave.tags;

// Java
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

// JSON
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

// Program
import jgappsandgames.smartreminderssave.tasks.Task;
import jgappsandgames.smartreminderssave.tasks.TaskManager;
import jgappsandgames.smartreminderssave.utility.JSONUtility;
import jgappsandgames.smartreminderssave.utility.API;

/**
 * TagManager
 * Created by joshua on 8/26/17.
 *
 * Currently on API: 10
 **/
public class TagManager {
	// Constants
	private static final String FILENAME = "tagmanager.srj";
	
	private static final String VERSION = "version";
	private static final String META = "meta";

	private static final String TAGS = "tags";
	
	// Data
	private static int version;
	public static JSONObject meta;

	public static ArrayList<String> tags;
	
	// Management Methods
	public static void create() {
		version = API.MANAGEMENT;
        tags = new ArrayList<>();

        // VERSION 11
        meta = new JSONObject();
	}

	public static void load() {
	    try {
            loadJSON(JSONUtility.loadJSON(new File(FileUtility.getApplicationDataDirectory(), FILENAME)));
        } catch (IOException i) {
	        i.printStackTrace();
	        create();
	        save();
        }
    }

    public static void save() {
        JSONUtility.saveJSONObject(new File(FileUtility.getApplicationDataDirectory(), FILENAME), toJSON());
    }

    // JSON Management Methods
    public static void loadJSON(JSONObject data) {
        if (data == null) {
            create();
            return;
        }

        version = data.optInt(VERSION, API.RELEASE);
        JSONArray t = data.optJSONArray(TAGS);
        tags = new ArrayList<>(t.length());
        for (int i = 0; i < t.length(); i++) tags.add(t.optString(i));

        // Version 11
        if (version >= API.MANAGEMENT) {
            meta = data.optJSONObject(META);
            if (meta == null) meta = new JSONObject();
        } else {
            meta = new JSONObject();
        }
    }

    public static JSONObject toJSON() {
        JSONObject data = new JSONObject();
        JSONArray t = new JSONArray();

        try {
            for (String tag : tags) t.put(tag);

            data.put(VERSION, API.MANAGEMENT);
            data.put(TAGS, t);
            data.put(META, meta);
        } catch (JSONException j) {
            j.printStackTrace();
        }

        return data;
    }

    public static void addMissingTags(JSONObject data) {
        JSONArray t = data.optJSONArray(TAGS);
        tags = new ArrayList<>(t.length());
        for (int i = 0; i < t.length(); i++) if (!tags.contains(t.optString(i))) tags.add(t.optString(i));
    }

    // TagManagement Methods
    public static boolean addTag(String tag) {
	    // Check to See if the Tag is equal to ""
        if (tag.equals("")) return false;

        // Check to See if the Tag is Already there
        if (tags.contains(tag)) return false;

        // Add the Tag
        tags.add(tag);
        return true;
    }

    public static void deleteTag(String tag) {
        // Check to See if the Tag is equal to ""
        if (tag.equals("")) return;

        // Check to See if the Tag is Already there
        if (!tags.contains(tag)) return;

        // Remove Tag
        tags.remove(tag);

        // Clear Tags from Tasks
        for (String t : TaskManager.tasks) {
            Task task = new Task(t);
            task.removeTag(tag).save();
        }
    }

    public static boolean contains(String tag) {
	    return tags.contains(tag);
    }
}

File: smartreminderssave/src/main/java/jgappsandgames/smartreminderssave/tags/TagManager.kt
package jgappsandgames.smartreminderssave.tags

// Java
import jgappsandgames.smartreminderssave.settings.getUseVersion
import java.io.File

// JSON
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

// Save Library
import jgappsandgames.smartreminderssave.utility.*

// Settings
import jgappsandgames.smartreminderssave.tasks.getAllTasks
import jgappsandgames.smartreminderssave.tasks.returnTaskToPool

/**
 * TagManager
 * Created by joshua on 12/12/2017.
 */
// Constants -------------------------------------------------------------------------------
private const val FILENAME = "tagmanager.srj"

private const val VERSION = "version"
private const val META = "meta"
private const val TAGS = "tags"

private const val VERSION_12 = "a"
private const val META_12 = "b"
private const val TAGS_12 = "c"

// Data ------------------------------------------------------------------------------------
private var version = MANAGEMENT
@JvmField var meta: JSONObject = JSONObject()
private val tags: ArrayList<String> = ArrayList()

// Management Methods ----------------------------------------------------------------------
fun createTags() {
    if (getApplicationDataFile(FILENAME).exists()) loadTags()

    version = RELEASE
    tags.clear()
    meta = JSONObject()

    saveTags()
}

fun loadTags() {
    val data = loadJSONObject(File(getApplicationDataDirectory(), FILENAME))
    version = data.optInt(VERSION, RELEASE)

    if (version <= MANAGEMENT) {
        val t = data.optJSONArray(TAGS)
        tags.clear()
        for (i in 0 until t.length()) tags.add(t.optString(i))

        // API 11
        meta = if (version >= MANAGEMENT) data.optJSONObject(META)
        else JSONObject()
    }

    else {
        val t = data.optJSONArray(TAGS_12)
        tags.clear()
        for (i in 0 until t.length()) tags.add(t.optString(i))
        meta = data.optJSONObject(META_12)
    }

    tags.sort()
    saveTags()
}

fun saveTags() {
    tags.sort()
    saveJSONObject(File(getApplicationDataDirectory(), FILENAME), toJSON())
}

private fun toJSON(): JSONObject {
    val data = JSONObject()
    val t = JSONArray()

    try {
        if (getUseVersion() <= MANAGEMENT) {
            for (tag in tags) t.put(tag)

            data.put(VERSION, MANAGEMENT)
            data.put(TAGS, t)

            // API 11
            data.put(META, meta)
        } else {
            for (tag in tags) t.put(tag)

            data.put(VERSION, SHRINKING)
            data.put(VERSION_12, SHRINKING)
            data.put(META_12, meta)
            data.put(TAGS_12, t)
        }
    } catch (j: JSONException) {
        j.printStackTrace()
    }

    return data
}

// Tag Management Methods ------------------------------------------------------------------
fun addTag(tag: String): Boolean {
    if (tag == "") return false

    return when {
        tags.size == 0 -> {
            tags.add(tag)
            tags.sort()
            saveTags()
            true
        }

        tags.contains(tag) -> false

        else -> {
            tags.add(tag)
            tags.sort()
            saveTags()
            true
        }
    }
}

fun deleteTag(tag: String) {
    if (tag == "") return
    if (!tags.contains(tag)) return

    tags.remove(tag)
    saveTags()

    val all = getAllTasks()
    for (t in all) {
        t.removeTag(tag).save()
        returnTaskToPool(t)
    }
}

fun getTags(): ArrayList<String> {
    return tags
}

fun contains(tag: String): Boolean = tags.contains(tag)